app/jobs/report_job.rb

```diff
@@ -1,15 +1,29 @@
 # frozen_string_literal: true
 
-# Renders a report and saves it to a temporary file.
+# Renders a report and stores it in a given blob.
 class ReportJob < ApplicationJob
-  def perform(report_class, user, params, format)
+  def self.create_blob!
+    # ActiveStorage discourages modifying a blob later but we need a blob
+    # before we know anything about the report file. It enables us to use the
+    # same blob in the controller to read the result.
+    ActiveStorage::Blob.create_before_direct_upload!(
+      filename: "tbd",
+      byte_size: 0,
+      checksum: "0",
+      content_type: "application/octet-stream",
+    ).tap do |blob|
+      ActiveStorage::PurgeJob.set(wait: 1.month).perform_later(blob)
+    end
+  end
+
+  def perform(report_class, user, params, format, blob)
     report = report_class.new(user, params, render: true)
     result = report.render_as(format)
```

Comment:
It would be cool if one day we could pass a stream to the report, so that it can write line-by-line rather than save the entire contents in a large variable...

Reply:
Me too! Imagine we could read in batches with `find_each` and process the report in a stream. That would save a lot of memory. But we need the whole lot for grouping and sorting at the moment.

Reply:
As an aside, I discovered yesterday what looks like an alternative report method, building an SQL query here: https://github.com/openfoodfoundation/openfoodnetwork/blob/master/lib/reporting/reports/enterprise_fee_summary/scope.rb

Reply:
Ah, right. And I remember Matt introducing a new gem which makes custom SQL queries a lot more convenient. That's probably the way to go.